fe/fe-core/src/main/java/com/starrocks/qe/scheduler/WorkerProvider.java

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.qe.scheduler;
+
+import com.starrocks.common.Reference;
+import com.starrocks.system.ComputeNode;
+import com.starrocks.system.SystemInfoService;
+import com.starrocks.thrift.TNetworkAddress;
+
+import java.util.Collection;
+import java.util.List;
+
+public interface WorkerProvider {
+    interface Factory {
+        /**
+         * Capture the available workers from {@code systemInfoService}, which are alive and not in the blacklist.
+         *
+         * @param systemInfoService   The service which provides all the backend nodes and compute nodes.
+         * @param preferComputeNode   Whether to prefer using compute nodes over backend nodes.
+         * @param numUsedComputeNodes The maximum number of used compute nodes.
+         */
+        WorkerProvider captureAvailableWorkers(SystemInfoService systemInfoService,
+                                               boolean preferComputeNode,
+                                               int numUsedComputeNodes);
+    }
+
+    /**
+     * Choose the specific backend node.
+     *
+     * @param backendID The ID of the backend node to choose.
+     * @return The address with the {@code ComputeNode#bePort} of the backend node.
+     * @throws SchedulerException if there is no available backend with {@code backendID}.
+     */
+    TNetworkAddress chooseBackend(Long backendID) throws SchedulerException;
+
+    /**
+     * Choose the next worker node.
+     *
+     * @param workerIdRef storing ID of the worker node to choose.
+     * @return The address with the {@code ComputeNode#bePort} of the worker node.
+     * @throws SchedulerException if there is no available backend with {@code backendID}.
+     */
+    TNetworkAddress chooseNextWorker(Reference<Long> workerIdRef) throws SchedulerException;
+
+    /**
+     * Choose all the available compute nodes.
+     *
+     * @return The address with the {@code ComputeNode#bePort} of the compute nodes to choose.
+     */
+    List<TNetworkAddress> chooseAllComputedNodes();
+
+    Collection<ComputeNode> getWorkers();
+
+    boolean isBackendAvailable(Long backendID);
+
+    void reportBackendNotFoundException() throws SchedulerException;
+
+    void reportWorkerNotFoundException() throws SchedulerException;
```

Comment:
It is very strange that the reportXXXX exception methods are the separate interface

Reply:
When choosing replication of a tablet, it find the available backend with the minimum used bytes.
If there isn't available backend, then `reportBackendNotFoundException`.

If there isn't `reportBackendNotFoundException`, how to implement this?

```java
                for (final TScanRangeLocation location : scanRangeLocations.getLocations()) {
                    if (!workerProvider.isBackendAvailable(location.getBackend_id())) {
                        continue;
                    }

                    Long assignedBytes = BackendSelector.findOrInsert(assignedRowCountPerHost, location.server, 0L);
                    if (assignedBytes < minAssignedBytes) {
                        minAssignedBytes = assignedBytes;
                        minLocation = location;
                    }
                }

                if (minLocation == null) {
                    workerProvider.reportBackendNotFoundException();
                }
```

Reply:
why not directly throw the Exception?

Reply:
IMO, this can provide better encapsulation, because available and non-available are encapsulated in the `WorkerProvider`, and only expose a report exception method.

If we directly throw the Exception here., then:
- The caller need know what exception should be threw, when there isn't available backend.
- `WorkerProvider` should provides more methods `getAllBackends` and `getAllComputeNodes`.
- There are lots of places throw this error, so we need duplicate the throw exception code. What's worse, when we need modify the exception logic (such as change the Exception type to throw), we must modify all these places.